Source/Core/MaxHolesLambdaLifter.cs

```diff
@@ -0,0 +1,392 @@
+            var replacements = new List<Expr>();
+            foreach (Expr arg in nodeArgs) {
+                replacements.AddRange(_templates[arg].GetReplacements());
+            }
+
+            _templates[node] = new TemplateWithBoundVariables(replacements);
+        } else {
+            // todo (MR) Could we avoid the casting? Could we encode in type system that all children of hole nodes are holes?
+            var newArgs = from arg in nodeArgs select ((TemplateNoBoundVariables) _templates[arg]).GetReplacement();
+            var llReplacementExpr =
+                new NAryExpr(node.tok, node.Fun, newArgs.ToList(), node.Immutable) {
+                    TypeParameters = node.TypeParameters, Type = node.Type ?? node.ShallowType
+                };
+            _templates[node] = new TemplateNoBoundVariables(llReplacementExpr);
+        }
+
+        return node;
+    }
+
+    public override Expr VisitIdentifierExpr(IdentifierExpr node) {
+        Contract.Requires(node != null);
+        if (_templates.ContainsKey(node)) return node;
+        base.VisitIdentifierExpr(node);
+        if (IsBound(node.Decl)) {
+            _templates[node] = new TemplateWithBoundVariables();
+        } else {
+            _templates[node] = _templates[node.Decl];
+        }
+
+        return node;
+    }
+
+    public override Expr VisitLiteralExpr(LiteralExpr node) {
+        Contract.Requires(node != null);
+        if (_templates.ContainsKey(node)) return node;
+        base.VisitLiteralExpr(node);
+        _templates[node] = new TemplateNoBoundVariables(node);
+        return node;
+    }
+
+    public override Expr VisitLetExpr(LetExpr node) {
+        Contract.Requires(node != null);
+        if (_templates.ContainsKey(node)) return node;
+        base.VisitLetExpr(node);
+        var bodyTemplate = _templates[node.Body];
+        var varBodies = node.Rhss;
+        if (bodyTemplate.ContainsBoundVariables() || varBodies.Any(body => _templates[body].ContainsBoundVariables())) {
+            var replacements = new List<Expr>();
+            foreach (Expr body in varBodies) {
+                replacements.AddRange(_templates[body].GetReplacements());
+            }
+
+            replacements.AddRange(bodyTemplate.GetReplacements());
+            _templates[node] = new TemplateWithBoundVariables(replacements);
+        }
+        else {
+            var newRhss = from arg in varBodies select ((TemplateNoBoundVariables) _templates[arg]).GetReplacement();
+            LambdaLiftingTemplate template = new TemplateNoBoundVariables(
+                new LetExpr(node.tok, node.Dummies, newRhss.ToList(),
+                    node.Attributes, // todo (MR) document says attributes should be dropped?
+                    ((TemplateNoBoundVariables) _templates[node.Body])
+                    .GetReplacement()));
+            _templates[node] = template;
+            // todo (MR) explicitly set type of let expr here?
```

Comment:
If the resulting expr is supposed to be resolved, then you'd need to set it yourself here. (How was this done in the previous lambda lifter?)

Reply:
The previous lambda lifter just replaced one type of variables (free variables) with another type of variables (bound variables), so it didn't need to do anything special for let expressions. I.e., it could reuse the original let expression and its type. I think it is better to set the type manually so I did that now.